File: tests/fixtures/v2-upload-flow/positive/UploadFlow.java
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

class UploadFlow {
    void direct(HttpServletRequest request, String content) throws Exception {
        Files.writeString(request.getPart("upload").getSubmittedFileName(), content);
    }

    void propagated(HttpServletRequest request, String content) throws Exception {
        String requested = request.getPart("upload").getSubmittedFileName();
        String alias = requested;
        Files.writeString(alias, content);
    }

    void canonicalized(HttpServletRequest request, String content) throws Exception {
        Path destination = Paths.get(request.getPart("upload").getSubmittedFileName()).normalize();
        Files.writeString(destination, content);
    }

    boolean validBasename(String filename) {
        return Paths.get(filename).getFileName().toString().equals(filename);
    }
}

File: tests/fixtures/v2-upload-flow/positive/UploadFlow.cs
class UploadFlow
{
    void Direct(dynamic Request, string content)
    {
        File.WriteAllText(Request.Form.Files[0].FileName, content);
    }

    void Propagated(dynamic Request, string content)
    {
        var requested = Request.Form.Files[0].FileName;
        var forwardedValue = requested;
        File.WriteAllText(forwardedValue, content);
    }

    void Canonicalized(dynamic Request, string content)
    {
        var destination = Path.GetFullPath(Request.Form.Files[0].FileName);
        File.WriteAllText(destination, content);
    }

    bool ValidBasename(string filename)
    {
        return Path.GetFileName(filename) == filename;
    }
}
